File: VB/DocumentVariablesExample/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraRichEdit.API.Native
#Region "#usings"
Imports DevExpress.XtraRichEdit
Imports DevExpress.Services
#End Region ' #usings
Imports DevExpress.XtraRichEdit.Commands

Namespace DocumentVariablesExample
	Partial Public Class Form1
		Inherits Form
		Private richEdit As RichEditControl

		Public Sub New()
			InitializeComponent()
			richEditControl1.LoadDocument("Docs\invitation.docx")
			richEditControl1.Options.MailMerge.DataSource = New SampleData()
			AddHandler richEditControl2.Document.CalculateDocumentVariable, AddressOf eventHandler_CalculateDocumentVariable
			Me.richEdit = richEditControl1
		End Sub

		Private Sub btnMailMerge_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnMailMerge.Click
			Dim myMergeOptions As MailMergeOptions = richEditControl1.Document.CreateMailMergeOptions()
			myMergeOptions.MergeMode = MergeMode.NewSection
			'myMergeOptions.FirstRecordIndex = 1;
			'myMergeOptions.LastRecordIndex = 3;
			Me.Cursor = Cursors.WaitCursor
			richEditControl1.Document.MailMerge(myMergeOptions, richEditControl2.Document)
			Me.Cursor = Cursors.Default
			xtraTabControl1.SelectedTabPageIndex = 1
			'richEditControl2.ActiveView.ZoomFactor = 0.4f;
		End Sub
#Region "#servicesubst"
		Private Sub richEditControl1_MailMergeStarted(ByVal sender As Object, ByVal e As MailMergeStartedEventArgs) Handles richEditControl1.MailMergeStarted
			richEditControl1.RemoveService(GetType(IProgressIndicationService))
			richEditControl1.AddService(GetType(IProgressIndicationService), New MyProgressIndicatorService(richEditControl1, Me.progressBarControl1))
		End Sub
  #End Region ' #servicesubst

		Private Sub richEditControl1_MailMergeFinished(ByVal sender As Object, ByVal e As MailMergeFinishedEventArgs) Handles richEditControl1.MailMergeFinished
			richEditControl1.RemoveService(GetType(IProgressIndicationService))
		End Sub

#Region "#mailmergerecordstarted"
		Private Sub richEditControl1_MailMergeRecordStarted(ByVal sender As Object, ByVal e As MailMergeRecordStartedEventArgs) Handles richEditControl1.MailMergeRecordStarted
			Dim _range As DocumentRange = e.RecordDocument.InsertText(e.RecordDocument.Range.Start, String.Format("Created on {0:G}" & Constants.vbLf + Constants.vbLf, DateTime.Now))
			Dim cp As CharacterProperties = e.RecordDocument.BeginUpdateCharacters(_range)
			cp.FontSize = 8
			cp.ForeColor = Color.Red
			cp.Hidden = True
			e.RecordDocument.EndUpdateCharacters(cp)
		End Sub
#End Region ' #mailmergerecordstarted

#Region "#mailmergerecordfinished"
		Private Sub richEditControl1_MailMergeRecordFinished(ByVal sender As Object, ByVal e As MailMergeRecordFinishedEventArgs) Handles richEditControl1.MailMergeRecordFinished
			e.RecordDocument.AppendDocumentContent("Docs\bungalow.docx", DocumentFormat.OpenXml)
		End Sub
#End Region ' #mailmergerecordfinished

#Region "#calculatedocumentvariable"
		Private Sub eventHandler_CalculateDocumentVariable(ByVal sender As Object, ByVal e As CalculateDocumentVariableEventArgs) Handles richEditControl1.CalculateDocumentVariable
			Dim location As String = e.Arguments(0).Value.ToString()

			Console.WriteLine(e.VariableName & " " & location)

			If (location.Trim() = String.Empty) OrElse (location.Contains("<")) Then
				e.Value = " "
				e.Handled = True
				Return
			End If

			Select Case e.VariableName
				Case "Weather"
					Dim conditions As New Conditions()
					conditions = Weather.GetCurrentConditions(location)
					e.Value = String.Format("Forecast for {0}: " & Constants.vbLf & "Conditions: {1}" & Constants.vbLf & "Temperature (C) :{2}" & Constants.vbLf & "Humidity: {3}" & Constants.vbLf & "Wind: {4}" & Constants.vbLf, conditions.City, conditions.Condition, conditions.TempC, conditions.Humidity, conditions.Wind)
				Case "Location"
					Dim loc() As GeoLocation = GeoLocation.GeocodeAddress(location)
					e.Value = String.Format(" {0}" & Constants.vbLf & "Latitude: {1}" & Constants.vbLf & "Longitude: {2}" & Constants.vbLf, loc(0).Address, loc(0).Latitude.ToString(), loc(0).Longitude.ToString())
			End Select
			e.Handled = True
		End Sub
#End Region ' #calculatedocumentvariable
		Private Sub xtraTabControl1_Selected(ByVal sender As Object, ByVal e As DevExpress.XtraTab.TabPageEventArgs) Handles xtraTabControl1.Selected
			Select Case e.PageIndex
				Case 0
					richEdit = richEditControl1
					Me.btnMailMerge.Enabled = True
				Case 1
					richEdit = richEditControl2
					Me.btnMailMerge.Enabled = False
			End Select
		End Sub

		Private Sub btn_ShowCodes_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btn_ShowCodes.CheckedChanged
			Dim doc As Document = richEdit.Document
			Dim showCodes As Boolean = btn_ShowCodes.Checked
			doc.BeginUpdate()
			For Each f As Field In doc.Fields
				f.ShowCodes = showCodes
			Next f
			doc.EndUpdate()
		End Sub

		Private Sub btn_ShowHiddenText_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles btn_ShowHiddenText.CheckedChanged
			If btn_ShowHiddenText.Checked Then
				richEdit.Options.FormattingMarkVisibility.HiddenText = RichEditFormattingMarkVisibility.Visible
			Else
				richEdit.Options.FormattingMarkVisibility.HiddenText = RichEditFormattingMarkVisibility.Hidden
			End If
		End Sub
	End Class
End Namespace

File: CS/DocumentVariablesExample/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraRichEdit.API.Native;
#region #usings
using DevExpress.XtraRichEdit;
using DevExpress.Services;
#endregion #usings
using DevExpress.XtraRichEdit.Commands;

namespace DocumentVariablesExample
{
    public partial class Form1 : Form
    {
        RichEditControl richEdit;

        public Form1()
        {
            InitializeComponent();
            richEditControl1.LoadDocument("Docs\\invitation.docx");
            richEditControl1.Options.MailMerge.DataSource = new SampleData();
            richEditControl2.Document.CalculateDocumentVariable += new CalculateDocumentVariableEventHandler(eventHandler_CalculateDocumentVariable);
            this.richEdit = richEditControl1;
        }

        private void btnMailMerge_Click(object sender, EventArgs e)
        {
            MailMergeOptions myMergeOptions = richEditControl1.Document.CreateMailMergeOptions();
            myMergeOptions.MergeMode = MergeMode.NewSection;
            //myMergeOptions.FirstRecordIndex = 1;
            //myMergeOptions.LastRecordIndex = 3;
            this.Cursor = Cursors.WaitCursor;
            richEditControl1.Document.MailMerge(myMergeOptions, richEditControl2.Document);
            this.Cursor = Cursors.Default;
            xtraTabControl1.SelectedTabPageIndex = 1;
            //richEditControl2.ActiveView.ZoomFactor = 0.4f;
        }
#region #servicesubst
        private void richEditControl1_MailMergeStarted(object sender, MailMergeStartedEventArgs e)
        {
            richEditControl1.RemoveService(typeof(IProgressIndicationService));
            richEditControl1.AddService(typeof(IProgressIndicationService),
                new MyProgressIndicatorService(richEditControl1, this.progressBarControl1));
        }
  #endregion #servicesubst

        private void richEditControl1_MailMergeFinished(object sender, MailMergeFinishedEventArgs e)
        {
            richEditControl1.RemoveService(typeof(IProgressIndicationService));
        }

#region #mailmergerecordstarted
        private void richEditControl1_MailMergeRecordStarted(object sender, MailMergeRecordStartedEventArgs e)
        {
            DocumentRange _range = e.RecordDocument.InsertText(e.RecordDocument.Range.Start, 
String.Format("Created on {0:G}\n\n", DateTime.Now));
            CharacterProperties cp = e.RecordDocument.BeginUpdateCharacters(_range);
            cp.FontSize = 8;
            cp.ForeColor = Color.Red;
            cp.Hidden = true;
            e.RecordDocument.EndUpdateCharacters(cp);
        }
#endregion #mailmergerecordstarted

#region #mailmergerecordfinished
        private void richEditControl1_MailMergeRecordFinished(object sender, MailMergeRecordFinishedEventArgs e)
        {
            e.RecordDocument.AppendDocumentContent("Docs\\bungalow.docx", DocumentFormat.OpenXml);
        }
#endregion #mailmergerecordfinished

#region #calculatedocumentvariable
        void eventHandler_CalculateDocumentVariable(object sender, CalculateDocumentVariableEventArgs e)
        {
            string location = e.Arguments[0].Value.ToString();

            Console.WriteLine(e.VariableName + " " + location);

            if ((location.Trim() == String.Empty) || (location.Contains("<"))) {
                e.Value = " ";
                e.Handled = true;
                return;
            }

            switch (e.VariableName)
            {
                case "Weather":
                    Conditions conditions = new Conditions();
                    conditions = Weather.GetCurrentConditions(location);
                    e.Value = String.Format("Forecast for {0}: \nConditions: {1}\nTemperature (C) :{2}\nHumidity: {3}\nWind: {4}\n",
                        conditions.City, conditions.Condition, conditions.TempC, conditions.Humidity, conditions.Wind);
                    break;
                case "Location":
                    GeoLocation[] loc = GeoLocation.GeocodeAddress(location);
                    e.Value = String.Format(" {0}\nLatitude: {1}\nLongitude: {2}\n",
                        loc[0].Address, loc[0].Latitude.ToString(), loc[0].Longitude.ToString());
                    break;
            }
            e.Handled = true;
        }
#endregion #calculatedocumentvariable
        private void xtraTabControl1_Selected(object sender, DevExpress.XtraTab.TabPageEventArgs e)
        {
            switch (e.PageIndex) {
                case 0:
                    richEdit = richEditControl1;
                    this.btnMailMerge.Enabled = true;
                    break;
                case 1:
                    richEdit = richEditControl2;
                    this.btnMailMerge.Enabled = false;
                    break;
            }
        }

        private void btn_ShowCodes_Click(object sender, EventArgs e)
        {
            Document doc = richEdit.Document;
            bool showCodes = btn_ShowCodes.Checked;
            doc.BeginUpdate();
            foreach (Field f in doc.Fields) f.ShowCodes = showCodes;
            doc.EndUpdate();
        }

        private void btn_ShowHiddenText_CheckedChanged(object sender, EventArgs e)
        {
            richEdit.Options.FormattingMarkVisibility.HiddenText = btn_ShowHiddenText.Checked ? RichEditFormattingMarkVisibility.Visible : RichEditFormattingMarkVisibility.Hidden;
        }
    }
}